### PR TITLE
build script; OS X + brew does not put libgc.a on OS linker path; extend it

### PR DESCRIPTION
**Pull request by pnkfelix**
(You should not accept this PR; at least, not yet.)

I saw in the [README.md](../blob/master/README.md) that you say that it works for you with homebrew, but for me I needed to get cargo to add linker directives to the `rustc` invocation.

Do you have some other means for getting around this?

Anyway, this build script is not yet robust enough for inclusion. (It should probably emit nothing if the file is not found, but not panic at this point.  Ideally we would perhaps also invoke a compile of a dummy .c program that needs to link to libgc and make sure that doing so succeeds.)

### CODE REVIEW

**Comment by swgillespie** [comment]
Thanks!

It's been a little while since I looked at this (i'll refresh/rebuild this when I get a chance later today) but as I recall, I did not need to pass any specific linker instructions to `rustc` after installing with homebrew. I'll go check this again, though - it's very possible something has changed in homebrew since I did this or I've changed on my machine and forgot about it.

When I first did this I also thought about finding and building libgc from source as part of the build script so that we can avoid this problem and also have control over the #defined feature flags, in particular the ones that influence how the GC behaves in multithreaded environments. I ran into problems building libgc from source on my machine so I stopped, but we could certainly try again.

**Comment by pnkfelix** [comment]
hmm, now I'm pretty sure that I needed this `build.rs` script to get things going yesterday ...

... but at the same time, I am now encountering behavior where it seems to be "successfully" linking to `/usr/local/lib/libgc.dylib` when I put in a build script that is deliberately trying to provide a _different_ directory to use for linking ...

... curiouser and curiouser...

**Comment by pnkfelix** [comment]
@swgillespie 

> When I first did this I also thought about finding and building libgc from source as part of the build script so that we can avoid this problem and also have control over the #defined feature flags, in particular the ones that influence how the GC behaves in multithreaded environments. I ran into problems building libgc from source on my machine so I stopped, but we could certainly try again.

I got a build script going that can run `configure` and `make` on a copy of the embedded source code that I put into the repo itself. (Which may imply licensing issues -- but you don't yet have a `LICENSE` file in the repo anyway...)

If you want to see how I hacked that up, you can look at it here:

  https://github.com/pnkfelix/boehm_gc_allocator/blob/build-own-libgc/build.rs

---

Update: nonetheless, I have not managed to get this to work with my smoketest without requiring the per-thread registration nonetheless.
1. It seems like `DARWIN_DONT_PARSE_STACK` may be defined for all darwin targets (despite the documentation implying that it is not the default).
2. But even after I manually got rid of that definition, things broke in worse ways (null ptr dereference within the GC code, I _think_, not sure, its late here...)

Basically, I think a local build may be a good way to avoid having to call the initial "gc_allow_register_threads" (as you said [here](https://github.com/swgillespie/boehm_gc_allocator/issues/2#issuecomment-154138720)), but so far I have not managed to get the old threads-discovery code working well enough for us to rely on, I think.

Still, maybe requiring the

``` rust
::std::thread::at_start(|| { unsafe { boehm_gc::gc_register_myself(); } });
```

without having to do the `gc_allow_register_threads` will be good enough. alex crichton did not seem immediately opposed to `::std::thread::at_start` when I mentioned it to him yesterday.

**Comment by swgillespie** [comment]
I also tried to get thread discovery working yesterday on OSX but I also had no luck - perhaps the entire feature is deprecated instead of just the API calls themselves?

At any rate, based on your experience and my own I'm afraid we might be hosed without `at_start`. I see the PreRFC you made on the discussion board and I'll participate in that discussion too.

I've been pondering over the past few days whether it would be possible to provide this as library functionality, but I can't think of anything that would avoid hooking into `thread::spawn`...
